tornadox/solvers/ek1.py

```diff
@@ -0,0 +1,192 @@
+from tornadox.blocks.inits import autodiff_first_order
+from tornadox.blocks.sde import ibm
+from tornadox.blocks.step_impl import ek1_projmatfree_d_nu_first_order
+
+
+def ek1_terminal_value(*, ode_dimension, num_derivatives=5):
+    """The traditional EK1, but only for computing the terminal value.
+
+    Uses adaptive steps and PI control.
+    Uses time-varying, scalar-valued diffusion.
+    Uses Taylor-mode initialisation for num_derivatives >= 5,
+    and forward-mode initialisation otherweise.
+
+    Uses a dimension-derivative ordering internally,
+    and does not use projection matrices.
+    """
+
+    EK1State = namedtuple("EK1State", ("u", "dt_proposed", "error_norm", "stats"))
+
+    # Create the identity matrix required for Kronecker-type things below.
+    eye_d = jnp.eye(ode_dimension)
+    a_1d, q_sqrtm_1d = ibm.system_matrices_1d(num_derivatives=num_derivatives)
+    a, q_sqrtm = jnp.kron(eye_d, a_1d), jnp.kron(eye_d, q_sqrtm_1d)
+
+    # Initialisation with autodiff
+    if num_derivatives <= 5:
+        autodiff_fun = autodiff_first_order.forwardmode_jvp
+    else:
+        autodiff_fun = autodiff_first_order.taylormode
+
+    @partial(jax.jit, static_argnames=("f", "df"))
+    def init_fn(
+        *,
+        f,
+        df,
+        tspan,
+        u0,
+        rtol,
+        atol,
+        safety=0.95,
+        factor_min=0.2,
+        factor_max=10.0,
+        power_integral_unscaled=0.3,
+        power_proportional_unscaled=0.4,
+    ):
+        del df  # unused
+
+        m0_mat = autodiff_fun(f=f, u0=u0, num_derivatives=num_derivatives)
+        m0 = m0_mat.reshape((-1,), order="F")
+        c_sqrtm0 = jnp.zeros((m0.size, m0.size))
+        dt0 = stepsizes.propose_first_dt_per_tol(
+            f=f, u0=u0, atol=atol, rtol=rtol, num_derivatives=num_derivatives
+        )
+
+        stats = {
+            "f_evaluation_count": 0,
+            "df_evaluation_count": 0,
+            "steps_accepted_count": 0,
+            "steps_attempted_count": 0,
+            "dt_min": jnp.inf,
+            "dt_max": 0.0,
+        }
+        state = EK1State(u=(m0, c_sqrtm0), dt_proposed=dt0, error_norm=1.0, stats=stats)
+        return tspan[0], state
+
+    @partial(jax.jit, static_argnames=("f", "df"))
+    def perform_step_fn(
+        t0,
+        state0,
+        *,
+        f,
+        df,
+        t1,
+        rtol,
+        atol,
+        safety=0.95,
+        factor_min=0.2,
+        factor_max=10.0,
+        power_integral_unscaled=0.3,
+        power_proportional_unscaled=0.4,
+    ):
+        """Perform a successful step."""
+
+        m0, c_sqrtm0 = state0.u
+        error_norm_previously_accepted = state0.error_norm
+
+        @jax.jit
+        def cond_fun(s):
+            _, state = s
+            return state.error_norm > 1
+
+        @jax.jit
+        def body_fun(s):
+            t, s_prev = s
+
+            # Never exceed the terminal value
+            dt_clipped = jnp.minimum(s_prev.dt_proposed, t1 - t0)
```

Comment:
Shouldn't this be
```suggestion
            dt_clipped = jnp.minimum(s_prev.dt_proposed, t1 - t)
```
?

Reply:
Ok actually maybe not, `t0` seems to not be what I think it is?

Reply:
t0 is the point of the previously accepted step.  `t` is actually unused (I made this fact more transparent, thanks for catching this).